### PR TITLE
CLDR-10564 Logical groups; unit test and refactoring

### PR DESCRIPTION
**Pull request by btangmu**
-New unit test TestCheckLogicalGroupings.testSameCoverageLevel

-Take isOptional into account for the unit test

-Refactor CheckLogicalGroupings.java, especially extremely long method handleCheck

-New LogicalGroupingChecker.java encapsulates the refactored code with shorter methods

CLDR-10564

- [ ] This PR completes the ticket.

### CODE REVIEW

File: tools/cldr-code/src/test/java/org/unicode/cldr/test/TestCheckLogicalGroupings.java [hunk 1]
```diff
@@ -79,4 +76,53 @@ private void assertCheckStatus(String localeID, CheckLogicalGroupings clg, Strin
...
+
+    void sameLevel(String localeId) {
+        final CLDRConfig config = CLDRConfig.getInstance();
+        final CLDRFile cldrFile = config.getCLDRFile(localeId, true);
+        final SupplementalDataInfo sdi = SupplementalDataInfo.getInstance();
+        final CoverageLevel2 coverageLevel = CoverageLevel2.getInstance(sdi, localeId);
+        for (final String path : cldrFile.fullIterable()) {
+            Set<String> grouping = LogicalGrouping.getPaths(cldrFile, path);
+            if (grouping != null && grouping.size() > 1) {
+                LogicalGrouping.removeOptionalPaths(grouping, cldrFile);
```
Comment (btangmu) [inline tools/cldr-code/src/test/java/org/unicode/cldr/test/TestCheckLogicalGroupings.java:99]: This "removeOptionalPaths" is the only difference between this PR and the preceding draft PR, as far as the unit test is concerned

**Review by macchiati — COMMENTED** [review]
Looks good. Let's hold off integrating until we are done vetting, unless you think there are good reasons to integrate.

File: tools/cldr-code/src/test/java/org/unicode/cldr/test/TestCheckLogicalGroupings.java (hunk 1 above)
Comment (srl295) [reply tools/cldr-code/src/test/java/org/unicode/cldr/test/TestCheckLogicalGroupings.java:99]: also the test "if (grouping != null) … " was added.

File: tools/cldr-code/src/main/java/org/unicode/cldr/test/CheckLogicalGroupings.java [hunk 2]
```diff
@@ -342,7 +145,7 @@ public int size() {
         }
 
         public static Fingerprint make(String value) {
-            return FINGERPRINT_CACHE.computeIfAbsent(value, v -> new Fingerprint(v));
+            return FINGERPRINT_CACHE.computeIfAbsent(value, Fingerprint::new);
```
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/test/CheckLogicalGroupings.java:148]: yes, this is the idiom i should be using elsewhere.

**Review by srl295 — APPROVED** [review]
lgtm
